### PR TITLE
Celestial trajectory plotting

### LINKED ISSUE #942: Draw celestial body orbits
@ferram4 strongly recommends we try to put this in Cantor, otherwise we'll have an endless stream of users confused by the strong wobbling of the primary-centred eccentricity (see also the end of Scott's video, and the e<sub>Bop</sub>(t) plot [here](https://goo.gl/photos/xXh83HTejv8vo8daA)---blue is KSP's eccentricity, yellow is the Jacobi eccentricity).

### LINKED ISSUE #2164: Support hiding histories instead of forgetting them
Many users set a shorten the history at some point because it gets in their way; since that setting forgets the history of all vessels, it makes it impossible to look back at any history again.

That setting should only hide the history, reversibly, and should probably be per-vessel instead of global.

Forgetting is more of a performance concern, and should be a separate setting.

### PR DESCRIPTION
**Pull request by eggrobin**
This pull request makes the `Max history length` setting hide history, instead of forgetting it. It does not introduce a separate setting to forget the history; we may want to add that if people complain about performance issues.

Fix #942.
Fix #2164.

All hail retrobop:
screenshot18

Epicycles:
screenshot20

Short histories:
screenshot21
screenshot22

Miscellaneous screenshots:
screenshot11
screenshot12
screenshot13
screenshot14
screenshot15
screenshot16
screenshot17

### CODE REVIEW

File: ksp_plugin/interface_planetarium.cpp [hunk 1]
```diff
@@ -203,12 +202,80 @@ Iterator* principia__PlanetariumPlotPsychohistory(
...
+}
+
+Iterator* principia__PlanetariumPlotCelestialTrajectoryForPsychohistory(
+    Planetarium const* const planetarium,
+    Plugin const* const plugin,
+    int const celestial_index,
+    char const* const vessel_guid,
+    double const max_history_length) {
+  journal::Method<journal::PlanetariumPlotCelestialTrajectoryForPsychohistory>
+      m({planetarium, plugin, celestial_index, vessel_guid, max_history_length});
```
Comment (pleroy) [inline ksp_plugin/interface_planetarium.cpp]: Parameters in column.

File: ksp_plugin/interface_planetarium.cpp [hunk 2]
```diff
@@ -189,11 +189,10 @@ Iterator* principia__PlanetariumPlotPsychohistory(
...
-  journal::Method<journal::PlanetariumPlotPsychohistory> m({planetarium,
-                                                            plugin,
-                                                            method,
-                                                            vessel_guid});
+    char const* const vessel_guid,
+    double const max_history_length) {
+  journal::Method<journal::PlanetariumPlotPsychohistory> m(
+      {planetarium, plugin, method, vessel_guid, max_history_length});
   CHECK_NOTNULL(plugin);
   CHECK_NOTNULL(planetarium);
```
Comment (pleroy) [inline ksp_plugin/interface_planetarium.cpp]: Should probably check `vessel_guid` too.

File: ksp_plugin/interface_planetarium.cpp [hunk 3]
```diff
@@ -203,12 +202,80 @@ Iterator* principia__PlanetariumPlotPsychohistory(
...
+  if (plugin->renderer().HasTargetVessel()) {
+    return m.Return(new TypedIterator<RP2Lines<Length, Camera>>({}));
+  } else {
+    auto const& celestial_trajectory =
+        plugin->GetCelestial(celestial_index).trajectory();
+    Instant const first_time = std::max(
+        plugin->CurrentTime() - max_history_length * Second,
+        vessel_guid == nullptr
+            ? celestial_trajectory.t_min()
+            : plugin->GetVessel(vessel_guid)->psychohistory().front().time);
```
Comment (pleroy) [inline ksp_plugin/interface_planetarium.cpp]: `t_min()`.

File: ksp_plugin/interface_planetarium.cpp [hunk 4]
```diff
@@ -203,12 +202,80 @@ Iterator* principia__PlanetariumPlotPsychohistory(
...
+                    psychohistory.LowerBound(plugin->CurrentTime() -
+                                             max_history_length * Second),
+                    psychohistory.end(),
+                    plugin->CurrentTime(),
+                    /*reverse=*/true);
+    return m.Return(new TypedIterator<RP2Lines<Length, Camera>>(rp2_lines));
+  }
+}
+
+Iterator* principia__PlanetariumPlotCelestialTrajectoryForPsychohistory(
```
Comment (pleroy) [inline ksp_plugin/interface_planetarium.cpp]: Could you comment what this method does?  I frankly have no idea...

File: ksp_plugin/interface_planetarium.cpp [hunk 5]
```diff
@@ -203,12 +202,80 @@ Iterator* principia__PlanetariumPlotPsychohistory(
...
+}
+
+Iterator* principia__PlanetariumPlotCelestialTrajectoryForPredictionOrFlightPlan(
+    Planetarium const* const planetarium,
+    Plugin const* const plugin,
+    int const celestial_index,
+    char const* const vessel_guid) {
+  journal::Method<
+      journal::PlanetariumPlotCelestialTrajectoryForPredictionOrFlightPlan>
+      m({planetarium, plugin, celestial_index, vessel_guid});
```
Comment (pleroy) [inline ksp_plugin/interface_planetarium.cpp]: In column.

File: ksp_plugin/interface_planetarium.cpp [hunk 6]
```diff
@@ -203,12 +202,80 @@ Iterator* principia__PlanetariumPlotPsychohistory(
...
+        planetarium->PlotMethod2(celestial_trajectory,
+                                 first_time,
+                                 /*last_time=*/plugin->CurrentTime(),
+                                 /*now=*/plugin->CurrentTime(),
+                                 /*reverse=*/true);
+    return m.Return(new TypedIterator<RP2Lines<Length, Camera>>(rp2_lines));
+  }
+}
+
+Iterator* principia__PlanetariumPlotCelestialTrajectoryForPredictionOrFlightPlan(
```
Comment (pleroy) [inline ksp_plugin/interface_planetarium.cpp]: Same, comment what this method does.

File: ksp_plugin/interface_planetarium.cpp [hunk 7]
```diff
@@ -203,12 +202,80 @@ Iterator* principia__PlanetariumPlotPsychohistory(
...
+      m({planetarium, plugin, celestial_index, vessel_guid});
+  CHECK_NOTNULL(plugin);
+  CHECK_NOTNULL(planetarium);
+
+  // Do not plot the past when there is a target vessel as it is misleading.
+  if (plugin->renderer().HasTargetVessel()) {
+    return m.Return(new TypedIterator<RP2Lines<Length, Camera>>({}));
+  } else {
+    auto const& vessel = *plugin->GetVessel(vessel_guid);
+    Instant const prediction_final_time = vessel.prediction().back().time;
```
Comment (pleroy) [inline ksp_plugin/interface_planetarium.cpp]: `t_max()`

File: ksp_plugin/planetarium.cpp [hunk 8]
```diff
@@ -138,21 +138,30 @@ RP2Lines<Length, Camera> Planetarium::PlotMethod2(
...
-  auto previous_time = reverse ? last_time : begin_time;
+  return PlotMethod2(trajectory, begin_time, last_time, now, reverse);
+}
+
+RP2Lines<Length, Camera> Planetarium::PlotMethod2(
+    Trajectory<Barycentric> const& trajectory,
+    Instant const& first_time,
+    Instant const& last_time,
+    Instant const& now,
+    bool reverse) const {
```
Comment (pleroy) [inline ksp_plugin/planetarium.cpp]: `const`

File: ksp_plugin/planetarium.hpp [hunk 9]
```diff
@@ -88,6 +89,13 @@ class Planetarium {
       Instant const& now,
       bool reverse) const;
 
+  RP2Lines<Length, Camera> PlotMethod2(
```
Comment (pleroy) [inline ksp_plugin/planetarium.hpp]: Comment both `PlotMethod1` methods (dunno why the existing one has no comment).

File: ksp_plugin_adapter/gl_lines.cs [hunk 10]
```diff
@@ -109,8 +109,9 @@ public enum Style {
               rp2_line_iterator.IteratorGetRP2LineXY());
           if (previous_rp2_point.HasValue) {
             if (style == Style.Faded) {
-              colour.a = 1 - (float)(4 * index) / (float)(5 * size);
-              UnityEngine.GL.Color(colour);
+              var faded_colour = colour;
+              faded_colour.a *= 1 - (float)(4 * index) / (float)(5 * size);
```
Comment (pleroy) [inline ksp_plugin_adapter/gl_lines.cs]: Comment this formula.  I'm not sure what the `*=` is doing.

File: ksp_plugin_adapter/ksp_plugin_adapter.cs [hunk 11]
```diff
@@ -1677,6 +1675,16 @@ private enum UnityLayers {
     }
   }
 
+  private void RemoveStockTrajectoriesIfNeeded(CelestialBody celestial) {
+    if (celestial.orbitDriver == null) {
+      return;
+    }
+    celestial.orbitDriver.Renderer.drawMode =
+        main_window_.display_patched_conics
```
Comment (pleroy) [inline ksp_plugin_adapter/ksp_plugin_adapter.cs]: Do we know why we still have that option?

File: ksp_plugin_adapter/ksp_plugin_adapter.cs [hunk 12]
```diff
@@ -1785,24 +1794,69 @@ private enum UnityLayers {
...
-        MapView.MapIsEnabled &&
-        plugin_.HasVessel(main_vessel_guid);
-    if (ready_to_draw_active_vessel_trajectory) {
+    if (MapView.MapIsEnabled) {
       XYZ sun_world_position = (XYZ)Planetarium.fetch.Sun.position;
       using (DisposablePlanetarium planetarium =
                 GLLines.NewPlanetarium(plugin_, sun_world_position)) {
         GLLines.Draw(() => {
+          // Celestial trajectories.
+          foreach (CelestialBody celestial in FlightGlobals.Bodies) {
```
Comment (pleroy) [inline ksp_plugin_adapter/ksp_plugin_adapter.cs]: This method is already very large and hard to follow.  Please move the celestial-related code to a separate method.

File: ksp_plugin/interface_planetarium.cpp [hunk 13]
```diff
@@ -189,11 +189,10 @@ Iterator* principia__PlanetariumPlotPsychohistory(
...
     int const method,
-    char const* const vessel_guid) {
-  journal::Method<journal::PlanetariumPlotPsychohistory> m({planetarium,
-                                                            plugin,
-                                                            method,
-                                                            vessel_guid});
+    char const* const vessel_guid,
+    double const max_history_length) {
+  journal::Method<journal::PlanetariumPlotPsychohistory> m(
+      {planetarium, plugin, method, vessel_guid, max_history_length});
```
[the remark this answers is not in the document]
Comment (eggrobin) [reply ksp_plugin/interface_planetarium.cpp]: > we generally put the parameters in a single column as it helps making sure that they match the profile of the method.

We don't do that where it doesn't fit to the right of `m`. Here it doesn't fit.

File: ksp_plugin/interface_planetarium.cpp (hunk 2 above)
Comment (eggrobin) [reply ksp_plugin/interface_planetarium.cpp]: We could do that everywhere in this file, but that's a job for a separate pull request.

File: ksp_plugin/interface_planetarium.cpp (hunk 1 above)
Comment (eggrobin) [reply ksp_plugin/interface_planetarium.cpp]: same.

File: ksp_plugin/interface_planetarium.cpp (hunk 2 above)
Comment (eggrobin) [reply ksp_plugin/interface_planetarium.cpp]: same.

File: ksp_plugin/interface_planetarium.cpp (hunk 5 above)
Comment (eggrobin) [reply ksp_plugin/interface_planetarium.cpp]: same.

File: ksp_plugin/planetarium.hpp (hunk 9 above)
Comment (eggrobin) [reply ksp_plugin/planetarium.hpp]: I think you mean 2.

File: ksp_plugin_adapter/ksp_plugin_adapter.cs (hunk 11 above)
Comment (eggrobin) [reply ksp_plugin_adapter/ksp_plugin_adapter.cs]: Mostly comparing with stock, for our entertainment.
